learn/sherif/main_test/src/main/java/frc/robot/subsystems/Factory.java

```diff
@@ -2,19 +2,19 @@
 
 import java.util.HashMap;
 
-import com.revrobotics.CANSparkMax;
 import com.revrobotics.CANSparkMaxLowLevel.MotorType;
 
-import edu.wpi.first.wpilibj.SpeedController;
+import frc.robot.CANSparkFunctions;
+import frc.robot.MockableSparkMax;
 
 public class Factory {
-    private HashMap<Integer, SpeedController> _map = new HashMap<>();
+    private HashMap<Integer, CANSparkFunctions> _map = new HashMap<>();
 
-    public SpeedController getMotor(int busID) {
+    public CANSparkFunctions getMotor(int busID) {
         if (this._map.containsKey(busID)) {
             return this._map.get(busID);
         }
-        var controller = new CANSparkMax(busID, MotorType.kBrushless);
+        var controller = new MockableSparkMax(busID, MotorType.kBrushless);
         this._map.put(busID, controller);
         return controller;
```

Comment:
I'm not totally sold on having the class maintain the map of controllers.  It seems like the requirement to have a single controller per bus id could also be satisfied by ensuring controllers are owned by subsystems.  This would be more explicit and would also enforce we don't have different parts of the code attempting to change the same controller.

That said, there is a better pattern for writing these kinds of functions that maintain a cache:
```
  if (!this._map.containsKey(busID)) {
    this._map.put(busID, new MockableSparkMax(busID, MotorType.kBrushless));
  }
  return this._map.get(busID);
```

This style ensure that the function always is returning the cached controller and there are no differences in behavior for the cached vs uncached cases.  In this code it doesn't matter too much, but when the initialization or caching code is more complicated then structuring the function that it conditionally populates the cache and then returns the cached value prevents bug.

Reply:
I understand the concern about having a cache and sharing components between subsystems. For now, I created a temporary solution by moving the cached factory into a special subclass `CachedFactory`. It's superclass `Factory` won't have the cache, just returning the controller directly.